org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SemanticHighlightingManager.xtend

```diff
@@ -0,0 +1,160 @@
+ * Copyright (c) 2018 TypeFox and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server
+
+import com.google.common.base.Preconditions
+import com.google.common.collect.BiMap
+import com.google.common.collect.ImmutableBiMap
+import com.google.common.collect.ImmutableList
+import com.google.common.primitives.Ints
+import com.google.inject.Singleton
+import java.util.List
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import org.eclipse.lsp4j.SemanticHighlightingInformation
+import org.eclipse.lsp4j.SemanticHighlightingParams
+import org.eclipse.lsp4j.VersionedTextDocumentIdentifier
+import org.eclipse.lsp4j.util.SemanticHighlightingTokens
+import org.eclipse.xtend.lib.annotations.Data
+import org.eclipse.xtext.util.internal.Log
+
+import static extension org.eclipse.lsp4j.util.SemanticHighlightingTokens.*
+
+/**
+ * Shared semantic highlighting manager per language server.
+ * Responsible for converting the semantic highlighted ranges into the LSP standard by producing a compact,
+ * {@code base64} encoded token string.
+ * 
+ * @since 2.15
+ */
+@Log
+@Singleton
+class SemanticHighlightingManager {
+
+	/**
+	 * Reserved TextMate scope identifier for styles that cannot be handled.
+	 */
+	public static val UNKNOWN_SCOPE = 'scope.unknown.xtext';
+
+	/**
+	 * TextMate scopes indicating an unhandled style ID to TextMate scope mapping.
+	 */
+	public static val UNKNOWN_SCOPES = #[UNKNOWN_SCOPE];
+
+	/**
+	 * A highlighted range with additional <a href="https://manual.macromates.com/en/language_grammars">TextMate scopes</a> information.
+	 * Note, multi-line highlighting ranges are not supported, hence the {@code length} instead of an {@link Range end range}.
+	 */
+	@Data
+	static class HighlightedRange extends Range {
+
+		/**
+		 * The internal index of the corresponding TextMate scope.
+		 */
+		val int scope;
+
+		/**
+		 * The length of the range.
+		 */
+		val int length;
+
+		new(Position start, int length, int scope) {
+			super(start, new Position(start.line, start.character + length));
+			this.scope = scope;
+			this.length = length;
+		}
+	}
+
+	/**
+	 * Lookup table for all known TextMate scopes.
+	 */
+	protected BiMap<Integer, List<String>> scopes;
+
+	/**
+	 * Initializes this manager with all known semantic highlighting TextMate scopes.
+	 */
+	def void initialize(List<List<String>> scopes) {
+		Preconditions.checkState(this.scopes === null, 'Already initialized.');
+		if (scopes === null) {
+			scopes = ImmutableBiMap.of;
+			return;
+		}
+		val builder = ImmutableBiMap.builder;
+		scopes.toSet.forEach [ scope, index |
```

Comment:
If the scopes should be a unique, why isn't it a Set in the first place?

Reply:
The order matters, I will check.